tactics/proof_global.ml

```diff
@@ -266,16 +163,120 @@ let return_proof ?(allow_partial=false) ps =
+  let nf = UnivSubst.nf_evars_and_universes_opt_subst subst_evar (UState.subst uctx) in
+
+  let make_entry (body, eff) (_, typ) =
+    let allow_deferred =
+      not poly && (keep_body_ucst_separate ||
+                   not (Safe_typing.empty_private_constants = eff.Evd.seff_private))
+    in
+    (* EJGA: Why are we doing things this way? *)
+    let typ = EConstr.Unsafe.to_constr typ in
+    let typ = if allow_deferred then typ else nf typ in
+    (* EJGA: End "Why are we doing things this way?" *)
 
-let close_proof ~opaque ~keep_body_ucst_separate fix_exn ps =
-  close_proof ~opaque ~keep_body_ucst_separate ~now:true
-    (Future.from_val ~fix_exn (return_proof ps)) ps
+    let used_univs_body = Vars.universes_of_constr body in
+    let used_univs_typ = Vars.universes_of_constr typ in
+    let used_univs = Univ.LSet.union used_univs_body used_univs_typ in
+    let utyp, ubody =
+      if allow_deferred then
+        let utyp = UState.univ_entry ~poly initial_euctx in
+        let uctx = UState.constrain_variables (fst (UState.context_set initial_euctx)) uctx in
+        (* For vi2vo compilation proofs are computed now but we need to
+           complement the univ constraints of the typ with the ones of
+           the body.  So we keep the two sets distinct. *)
+        let uctx_body = UState.restrict uctx used_univs in
+        let ubody = UState.check_mono_univ_decl uctx_body udecl in
+        utyp, ubody
+      else if poly && opaque && private_poly_univs () then
+        let universes = UState.restrict uctx used_univs in
+        let typus = UState.restrict universes used_univs_typ in
+        let utyp = UState.check_univ_decl ~poly typus udecl in
+        let ubody = Univ.ContextSet.diff
+            (UState.context_set universes)
+            (UState.context_set typus)
+        in
+        utyp, ubody
+      else
+        (* Since the proof is computed now, we can simply have 1 set of
+           constraints in which we merge the ones for the body and the ones
+           for the typ. We recheck the declaration after restricting with
+           the actually used universes.
+           TODO: check if restrict is really necessary now. *)
+        let ctx = UState.restrict uctx used_univs in
+        let utyp = UState.check_univ_decl ~poly ctx udecl in
+        utyp, Univ.ContextSet.empty
+    in
+    Declare.definition_entry ~opaque ?section_vars ~univs:utyp ~univc:ubody ~types:typ ~eff body
+  in
+  let entries = CList.map2 make_entry elist (Proofview.initial_goals entry) in
+  { name; entries; uctx; udecl }
+
+let close_proof_delayed ~opaque ~keep_body_ucst_separate ?feedback_id
+                (fpl : closed_proof_output Future.computation) ps =
+  let { section_vars; proof; udecl; initial_euctx } = ps in
+  let { Proof.name; poly; entry; sigma } = Proof.data proof in
+  (* We don't allow poly = true in this path *)
+  assert(poly=false);
+
+  let opaque = match opaque with Opaque -> true | Transparent -> false in
```

Comment:
This is quite fishy, I don't know when this flag started to be passed to delayed_definition, but we can't delay transparent definition so something went wrong at some point.

Reply:
It's been there since the beginning of this code I think; the reason is that indeed two very different paths where handled by the same entry point, this is what the PR tries to address.

Note that there is one codepath in the STM which could call this function with `opaque = Transparent` thus we cannot add an assert here.

Reply:
Can you point it out? It is clearly a bug no?

Reply:
Code is here https://github.com/coq/coq/blob/5f5f9520ccf0f107d381e5874a3743f47e37c409/stm/stm.ml#L2488

This will hit the delayed `close_proof` paht. I'm afraid I don't understand that part of the stm enough as to judge whether that's a bug or not.

Reply:
But indeed I agree that in principle we should ensure delayed can only happen for opaque; my internal tree is looking pretty good I'd dare to say in terms of invariants; but it is way too ahead of master unfortunately.

Reply:
Already @ppedrot enforced different entries for opaque/const at the kernel level, it should be a matter of a few more PRs to push the cast up to the proof layer / clients.

Reply:
I guess collect will prevent this case, so maybe that line in the STM should be an `assert(false)` , or it could use improved typing / data structure to keep track of the impossible cases.

Let's try that and see.

Reply:
OK commit pushed.

Reply:
Looks better now